### PR TITLE
ENH: Allow size=0 in numpy.random.choice

### LINKED ISSUE #8311: Allow random.choice from empty array when size=0
This piece of code:

```python
np.random.choice([],size=0)
```

Generates this error:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "mtrand.pyx", line 1397, in mtrand.RandomState.choice (numpy/random/mtrand/mtrand.c:15499)
ValueError: a must be non-empty
```

This error makes sense when `size>0`, but when `size=0`, `random.choice` could simply return `[]`. This would be useful when the `a` and `size` parameters are variable. Right now this situation requires an additional check:

```python
if size > 0:
  result = np.random.choice(a,size)
else:
  result = np.array([])
```

Would it be a good idea to simply return an empty array when `size=0`? This is my first issue so I hope this is the right way to address this.

I created a simple patch that implements this: 57ab0728d4cf73c1f4b3af2450fe603bf8670b75

I'm using Python 3.5.2 with numpy 1.11.2.

Edit: My implementation returns an empty slice from the original array. My thinking was that this would assign the correct `base` to the resulting array, but on second thought this is not what happens with `random.choice` normally, so I think returning `np.array([])` would be better.

### PR DESCRIPTION
**Pull request by mattip**
Replaces lost PR #8717. Fixes #8311. Includes tests.

@MareinK if you wish to continue I can give you permissions to push to the source repo for the PR.

### CODE REVIEW

**Comment by MareinK** [comment]
Thanks! Since you solved the conflict, I'm not sure what else needs to be done?

**Comment by mattip** [comment]
@eric-weiser approved the PR that this replaces, but requested a review from @seberg.

File: numpy/random/tests/test_random.py [hunk 1]
```diff
@@ -440,6 +440,14 @@ def test_choice_return_shape(self):
         assert_equal(np.random.choice(6, s, replace=False, p=p).shape, s)
         assert_equal(np.random.choice(np.arange(6), s, replace=True).shape, s)
 
+        # Check zero-size
+        assert_equal(np.random.randint(0,0,(3,0,4)).shape, (3,0,4))
+        assert_equal(np.random.randint(0,-10,0).shape, (0,))
+        assert_equal(np.random.choice(0,0).shape, (0,))
+        assert_equal(np.random.choice([],(0,)).shape, (0,))
```
Comment (eric-wieser) [inline numpy/random/tests/test_random.py]: nit: these tests would be clearer if the size/shape argument was passed by kwarg - too many zeros here

File: numpy/random/mtrand/mtrand.pyx [hunk 2]
```diff
@@ -1114,15 +1114,15 @@ cdef class RandomState:
...
+                raise ValueError("'a' must be greater than 0 unless no samples are taken")
         elif a.ndim != 1:
-            raise ValueError("a must be 1-dimensional")
+            raise ValueError("'a' must be 1-dimensional")
         else:
             pop_size = a.shape[0]
-            if pop_size is 0:
-                raise ValueError("a must be non-empty")
+            if pop_size is 0 and np.prod(size) != 0:
+                raise ValueError("a cannot be empty unless no samples are taken")
```
Comment (eric-wieser) [inline numpy/random/mtrand/mtrand.pyx]: If quotes are being added to all the `a`s, one is missed here

File: doc/release/1.16.0-notes.rst [hunk 3]
```diff
@@ -34,6 +34,12 @@ New Features
 Improvements
 ============
 
+``randint`` and ``choice`` now work on empty distributions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Even when no elements needed to be drawn, ``np.random.randint`` and
+``np.random.choice`` raised an error when the arguments described an empty
+distribution. This has been fixed so that e.g.
+``np.random.choice([],0) == np.array([],dtype=float64)``.
```
Comment (eric-wieser) [inline doc/release/1.16.0-notes.rst]: nit: spaces after commas

File: numpy/random/mtrand/mtrand.pyx (hunk 2 above)
Comment (mattip) [reply numpy/random/mtrand/mtrand.pyx]: fixed

File: numpy/random/tests/test_random.py (hunk 1 above)
Comment (mattip) [reply numpy/random/tests/test_random.py]: fixed

File: doc/release/1.16.0-notes.rst (hunk 3 above)
Comment (mattip) [reply doc/release/1.16.0-notes.rst]: fixed, also replaced all spaces after comments in these three files `grep ",[^ ]"`

File: numpy/random/tests/test_random.py [hunk 4]
```diff
@@ -1515,8 +1523,8 @@ def setup(self):
     def check_function(self, function, sz):
         from threading import Thread
 
-        out1 = np.empty((len(self.seeds),) + sz)
-        out2 = np.empty((len(self.seeds),) + sz)
+        out1 = np.empty((len(self.seeds), ) + sz)
+        out2 = np.empty((len(self.seeds), ) + sz)
```
Comment (eric-wieser) [inline numpy/random/tests/test_random.py]: I guess `(x,)` is an exception to the space after comma rule. Let me clarify that to "space after comma not before a closing group, like `]` or `)`"

Comment (mattip) [reply numpy/random/tests/test_random.py]: Fixed.
Shows that I should check [the standard](https://www.python.org/dev/peps/pep-0008/#whitespace-in-expressions-and-statements) before deciding matters of taste preference.

**Comment by eric-wieser** [comment]
Ideally all of the docstring whitespace changes would just go in a separate `STY` commit (or even PR), rather than muddying the blame.

File: numpy/random/tests/test_random.py [hunk 5]
```diff
@@ -440,6 +440,14 @@ def test_choice_return_shape(self):
         assert_equal(np.random.choice(6, s, replace=False, p=p).shape, s)
         assert_equal(np.random.choice(np.arange(6), s, replace=True).shape, s)
 
+        # Check zero-size
+        assert_equal(np.random.randint(0, 0, size=(3, 0, 4)).shape, (3, 0, 4))
```
Comment (eric-wieser) [inline numpy/random/tests/test_random.py]: A test for `randint(10, 10, size=0)` might be good too - just to check that we didn't special case 0 somehow

Comment (mattip) [reply numpy/random/tests/test_random.py]: done

**Review by eric-wieser — APPROVED** [review]

File: numpy/random/mtrand/mtrand.pyx [hunk 6]
```diff
@@ -1114,15 +1114,15 @@ cdef class RandomState:
                 # __index__ must return an integer by python rules.
                 pop_size = operator.index(a.item())
             except TypeError:
-                raise ValueError("a must be 1-dimensional or an integer")
-            if pop_size <= 0:
-                raise ValueError("a must be greater than 0")
+                raise ValueError("'a' must be 1-dimensional or an integer")
+            if pop_size <= 0 and np.prod(size) != 0:
```
Comment (seberg) [inline numpy/random/mtrand/mtrand.pyx]: a bit unintuitive that this works for `None`, but I guess OK, could also add `size is None` explicitly.

File: numpy/random/mtrand/mtrand.pyx [hunk 7]
```diff
@@ -1114,15 +1114,15 @@ cdef class RandomState:
...
+                raise ValueError("'a' must be greater than 0 unless no samples are taken")
```
Comment (seberg) [inline numpy/random/mtrand/mtrand.pyx]: I somewhat thought it was backticks ;). This is good, I do not think we or python has serious guidelines for errors.

File: numpy/random/mtrand/mtrand.pyx [hunk 8]
```diff
@@ -1136,9 +1136,9 @@ cdef class RandomState:
...
 
             if p.ndim != 1:
-                raise ValueError("p must be 1-dimensional")
+                raise ValueError("'p' must be 1-dimensional")
             if p.size != pop_size:
-                raise ValueError("a and p must have same size")
+                raise ValueError("'a' and 'p' must have same size")
             if np.logical_or.reduce(p < 0):
                 raise ValueError("probabilities are not non-negative")
             if abs(kahan_sum(pix, d) - 1.) > atol:
```
Comment (seberg) [inline numpy/random/mtrand/mtrand.pyx]: I think if you test for an empty probabilities array, you will see that this check fails also, so might as well allow that too?

File: numpy/random/mtrand/mtrand.pyx [hunk 9]
```diff
@@ -988,9 +988,9 @@ cdef class RandomState:
             raise ValueError("low is out of bounds for %s" % dtype)
         if ihigh > highbnd:
             raise ValueError("high is out of bounds for %s" % dtype)
-        if ilow >= ihigh:
-            raise ValueError("low >= high")
-
+        if ilow >= ihigh and np.prod(size) != 0:
+            raise ValueError("Range cannot be empty (low >= high) unless no samples are taken")
```
Comment (seberg) [inline numpy/random/mtrand/mtrand.pyx]: Was suprised for a bit here, but I guess we do it like a python `range` and allow strange ranges as empty ranges, seems fine to me.

File: numpy/random/mtrand/mtrand.pyx (hunk 6 above)
Comment (seberg) [reply numpy/random/mtrand/mtrand.pyx]: That also has the advantage that `np.prod` probably adds a bit of overhead, but maybe negligible in any case.

**Review by seberg — COMMENTED** [review]
Seems OK to me, allowing empty probabilities might be nice.

**Comment by mattip** [comment]
IMO handling 'p' should be a separate PR, see issues #11250, #9867, #6132, and PR #11264

**Comment by seberg** [comment]
OK, fine with me, thought it might have been an oversight.

File: numpy/random/mtrand/mtrand.pyx [hunk 10]
```diff
@@ -1114,15 +1114,15 @@ cdef class RandomState:
...
             except TypeError:
-                raise ValueError("a must be 1-dimensional or an integer")
-            if pop_size <= 0:
-                raise ValueError("a must be greater than 0")
+                raise ValueError("'a' must be 1-dimensional or an integer")
+            if pop_size <= 0 and np.prod(size) != 0:
+                raise ValueError("'a' must be greater than 0 unless no samples are taken")
         elif a.ndim != 1:
-            raise ValueError("a must be 1-dimensional")
+            raise ValueError("'a' must be 1-dimensional")
```
Comment (bashtage) [inline numpy/random/mtrand/mtrand.pyx]: I don't think it is normal to escape argument names in error messages, see, e.g., 

https://github.com/numpy/numpy/blob/master/numpy/random/mtrand/mtrand.pyx#L988

Comment (seberg) [reply numpy/random/mtrand/mtrand.pyx]: My guess is, you can find examples for everything both in numpy and the standard lib. Personally, I think quotes probably make it slightly more discoverable what `a` refers to, so sounds good to me. (e.g. something says `axis is invalid` may not refer to an axis argument, something saying `` `axis` is invalid`` certainly does).

Anyway, I think the PR looks good and you guys can put it in if you like.

**Comment by MareinK** [comment]
Since this issue and PR came from me originally: thanks to all who helped complete it!

**Pull request merged**